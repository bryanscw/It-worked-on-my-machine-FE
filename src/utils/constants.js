export const API_URL = 'http://localhost:3007';

export const USER_TYPES = {
    STUDENT: 'STUDENT',
    TEACHER: 'TEACHER',
    ADMIN: 'ADMIN',
};

export const CREATE = 'CREATE';
export const UPDATE = 'UPDATE';
export const DELETE = 'DELETE';
export const EMPTY = 'EMPTY';