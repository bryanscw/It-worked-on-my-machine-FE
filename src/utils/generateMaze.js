import generator from 'generate-maze';

export const EMPTY = 0;
export const WALL = 1;
export const GOAL = 2;

function shuffle(array) {
    var currentIndex = array.length, temporaryValue, randomIndex;
  
    // While there remain elements to shuffle...
    while (0 !== currentIndex) {
  
      // Pick a remaining element...
      randomIndex = Math.floor(Math.random() * currentIndex);
      currentIndex -= 1;
  
      // And swap it with the current element.
      temporaryValue = array[currentIndex];
      array[currentIndex] = array[randomIndex];
      array[randomIndex] = temporaryValue;
    }
  
    return array;
}

// Generates a maze of size 29 by 15
export default function generateMaze(numQuestions) {
    const WALL = 1;
    const EMPTY = 0;

    let originalMaze = generator(15, 8);
    let mazeWithWalls = [];

    for (let i = 0; i < originalMaze.length; i++) {
        let originalRow = originalMaze[i];
        let rowWithWalls = [];
        let rowBeneath = [];

        for (let j = 0; j < originalRow.length; j++) {
            rowWithWalls.push(EMPTY);
            rowWithWalls.push(originalRow[j].right ? WALL : EMPTY);
        }

        rowWithWalls.pop()

        for (let j = 0; j < originalRow.length; j++) {
            rowBeneath.push(originalRow[j].bottom ? WALL : EMPTY);
            rowBeneath.push(WALL);
        }

        rowBeneath.pop()

        mazeWithWalls.push(rowWithWalls);
        mazeWithWalls.push(rowBeneath);
    }
    
    mazeWithWalls.pop();

    let lastRow = mazeWithWalls[mazeWithWalls.length - 1];
    mazeWithWalls[mazeWithWalls.length - 1][lastRow.length - 1] = GOAL;

    let freeCoordinates = [];

    for (let i = 0; i < mazeWithWalls.length; i++) {
        let row = mazeWithWalls[i];

        for (let j = 0; j < row.length; j++) {
            if (row[j] === EMPTY && !(i === 0 && j === 0))
                freeCoordinates.push([i, j]);
        }
    }

    freeCoordinates = shuffle(freeCoordinates);
    const questionCoordinates = freeCoordinates.splice(0, numQuestions);
    let mapDescriptor = JSON.stringify(mazeWithWalls);

    return {
        mapDescriptor,
        questionCoordinates,
    };
}