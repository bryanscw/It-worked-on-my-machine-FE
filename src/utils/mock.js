import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

const authReducer = {
    user: {
        email: "test@test.com"
    },
    access_token: "123",
}

export const withRedux = (component, initialState = {}) => {
    const store = createStore(
        (state = {}, action) => state,
        {
            ...initialState,
            authReducer
        },
        applyMiddleware(thunk),
    )

    return <Provider store={store}>{component}</Provider>;
}

export const withRouter = (component, route = '/') => {
    const history = createMemoryHistory({ initialEntries: [route] });
    return <Router history={history}>{component}</Router>;
}

export const withReduxRouter = (component, initialState = {}, route = '/') => {
    const store = createStore(
        (state = {}, action) => state,
        {
            ...initialState,
            authReducer
        },
        applyMiddleware(thunk),
    )
    const history = createMemoryHistory({ initialEntries: [route] });

    return (
        <Provider store={store}>
            <Router history={history}>{component}</Router>
        </Provider>
    );
}