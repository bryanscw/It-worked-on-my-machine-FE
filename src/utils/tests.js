import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

import rootReducer from '../redux/rootReducer';

export const renderWithRedux = (component, initialState = {}) => {
    const store = createStore(
        rootReducer,
        initialState,
        applyMiddleware(thunk),
    )
    return <Provider store={store}>{component}</Provider>;
}

export const renderWithRouter = (component, route = '/') => {
    const history = createMemoryHistory({ initialEntries: [route] });
    return <Router history={history}>{component}</Router>
}

export const renderWithReduxRouter = (component, initialState = {}, route = '/') => {
    return renderWithRedux(renderWithRouter(component, route), initialState);
}