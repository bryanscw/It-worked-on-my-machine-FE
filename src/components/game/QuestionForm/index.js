import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { EMPTY } from '../../../utils/constants';

/**
 * This component displays the Question interface in the Popup
 */

export class QuestionForm extends Component {
    emptyState = {
        answer: null,
    }
    state = this.emptyState

    componentDidMount() {
        const {
            initialState
        } = this.props;
        
        if (initialState) {
            if (initialState === EMPTY)
                this.setState(this.emptyState);

            else
                this.setState(initialState);
        }
    }

    componentDidUpdate(prevProps) {
        const {
            initialState
        } = this.props;

        if (initialState && initialState !== prevProps.initialState) {
            if (initialState === EMPTY)
                this.setState(this.emptyState);
                
            else
                this.setState(initialState);
        }
    }

    onSubmit = e => {
        e.preventDefault();

        const {
            answer,
        } = this.state;

        this.props.onSubmit(answer);

        this.setState({
            answer: null
        });
    }

    onChangeAnswer = answer => {
        this.setState({
            answer: answer
        });
    }

    render() {
        const {
            options
        } = this.props;

        const {
            answer,
        } = this.state;

        return (
            <form className="form" onSubmit={this.onSubmit}>
                <p><strong>Options</strong></p>
                {
                    Object.keys(options).map(key => {
                        let option = options[key];
                        
                        return (
                            <div className="form-check" key={key}>
                                <input
                                    className="form-check-input"
                                    id={`option-${key}`}
                                    type="radio"
                                    name="radio"
                                    onChange={e => this.onChangeAnswer(e.target.checked ? key : null)}
                                    checked={answer === key}
                                    />
                                <label className="form-check-label" htmlFor={`option-${key}`}>{option}</label>
                            </div>
                        )
                    })
                }
                <div className="form-group mt-4">
                    <button className="btn btn-primary" type="submit">Submit</button>
                </div>
            </form>
        )
    }
}

QuestionForm.propTypes = {
    /** The initial state of the form. If it's 'EMPTY', then the form will be empty. If it's an object, the form will display the form with initial state's content. */
    initialState: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    /** The function to call when the form is submitted */
    onSubmit: PropTypes.func.isRequired,
}

export default QuestionForm
