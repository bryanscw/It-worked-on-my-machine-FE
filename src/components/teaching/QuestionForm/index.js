import React, { Component } from 'react';
import { EMPTY } from '../../../utils/constants';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faPlus, faTrash } from '@fortawesome/free-solid-svg-icons';

export class QuestionForm extends Component {
    emptyState = {
        questionText: '',
        options: [],
        answer: null,
    };
    state = this.emptyState;

    componentDidMount() {
        const {
            initialState
        } = this.props;

        if (initialState) {
            if (initialState === EMPTY)
                this.setState(this.emptyState);

            else
                this.setState(initialState);
        }
    }

    componentDidUpdate(prevProps) {
        const {
            initialState
        } = this.props;

        if (initialState && initialState !== prevProps.initialState) {
            if (initialState === EMPTY)
                this.setState(this.emptyState);
                
            else
                this.setState(initialState);
        }
    }

    onSubmit = e => {
        e.preventDefault();

        const {
            questionText,
            options,
            answer,
        } = this.state;

        this.props.onSubmit({
            questionText,
            options,
            answer,
        });

        this.setState(this.emptyState);
    }

    onChange = e => {
        this.setState({
            [e.target.name]: e.target.value
        });
    }

    onChangeOption = (selectedIndex, value) => {
        this.setState({
            options: this.state.options.map((option, i) => i === selectedIndex ? value : option),
        });
    }

    onChangeAnswer = answer => {
        this.setState({
            answer: answer
        });
    }

    onCreateOption = () => {
        this.setState({
            options: [...this.state.options, ""]
        });
    }

    onDeleteOption = selectedIndex => {
        let answer = this.state.answer;

        if (selectedIndex < answer) {
            answer--;
        } else if (selectedIndex === answer) {
            answer = null;
        }

        this.setState({
            options: this.state.options.filter((option, i) => i !== selectedIndex),
            answer: answer,
        });
    }

    render() {
        const {
            questionText,
            options,
            answer,
        } = this.state;

        return (
            <form className="form" onSubmit={this.onSubmit}>
                <div className="form-group">
                    <label>Question Text</label>
                    <input
                        className="form-control"
                        type="text"
                        name="questionText"
                        onChange={this.onChange}
                        value={questionText}
                        />
                </div>
                <p><strong>Options (Select the Correct Option)</strong></p>
                {
                    options.map((option, i) => 
                        <div className="form-group d-flex" key={i}>
                            <div className="form-check mr-1">
                                <input
                                    className="form-check-input"
                                    type="radio"
                                    name="radio"
                                    onChange={e => this.onChangeAnswer(e.target.checked ? i : null)}
                                    checked={answer === parseInt(i)}
                                    />
                            </div>
                            <input
                                className="form-control w-100 mr-2"
                                type="text"
                                name="option"
                                onChange={e => this.onChangeOption(i, e.target.value)}
                                value={option}
                                />
                            <button className="btn btn-danger btn-sm" type="button" onClick={() => this.onDeleteOption(i)}>
                                <FontAwesomeIcon icon={faTrash}/>
                            </button>
                        </div>
                    )
                }
                <button className="btn btn-primary" type="button" onClick={() => this.onCreateOption()}>
                    <FontAwesomeIcon icon={faPlus} className="mr-2" />Add Option
                </button>
                <div className="form-group mt-4">
                    <button className="btn btn-primary" type="submit">Submit</button>
                </div>
            </form>
        )
    }
}

export default QuestionForm
