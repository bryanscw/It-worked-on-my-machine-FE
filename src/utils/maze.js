import generator from 'generate-maze';

import { TILE_MAPPING, MAP_HEIGHT, MAP_WIDTH } from './constants';

export const EMPTY = 0;
export const WALL = 1;
export const GOAL = 2;
export const QUESTION = 3;

function shuffle(array) {
    var currentIndex = array.length, temporaryValue, randomIndex;
  
    // While there remain elements to shuffle...
    while (0 !== currentIndex) {
  
      // Pick a remaining element...
      randomIndex = Math.floor(Math.random() * currentIndex);
      currentIndex -= 1;
  
      // And swap it with the current element.
      temporaryValue = array[currentIndex];
      array[currentIndex] = array[randomIndex];
      array[randomIndex] = temporaryValue;
    }
  
    return array;
}

// Generates a maze of size 29 by 15
export function generateMaze(numQuestions) {
    let originalMaze = generator((MAP_WIDTH - 1) / 2, (MAP_HEIGHT - 1) / 2);
    let mazeWithWalls = [];

    for (let i = 0; i < originalMaze.length; i++) {
        let originalRow = originalMaze[i];
        let rowWithWalls = [];
        let rowBeneath = [];

        for (let j = 0; j < originalRow.length; j++) {
            rowWithWalls.push(EMPTY);
            rowWithWalls.push(originalRow[j].right ? WALL : EMPTY);
        }

        rowWithWalls.pop()

        for (let j = 0; j < originalRow.length; j++) {
            rowBeneath.push(originalRow[j].bottom ? WALL : EMPTY);
            rowBeneath.push(WALL);
        }

        rowBeneath.pop()

        mazeWithWalls.push(rowWithWalls);
        mazeWithWalls.push(rowBeneath);
    }
    
    mazeWithWalls.pop();

    // Add goal
    let lastRow = mazeWithWalls[mazeWithWalls.length - 1];
    mazeWithWalls[mazeWithWalls.length - 1][lastRow.length - 1] = GOAL;

    // Add surrounding walls
    let mazeWithSurroundings = mazeWithWalls.map(row => [WALL, ...row, WALL]);
    console.log(mazeWithSurroundings);

    let wallRow = [];
    for (let i = 0; i < MAP_WIDTH; i++)
        wallRow.push(WALL);

    mazeWithSurroundings = [wallRow, ...mazeWithSurroundings, wallRow];

    // Find free coordinates
    let freeCoordinates = [];

    for (let i = 0; i < mazeWithSurroundings.length; i++) {
        let row = mazeWithSurroundings[i];

        for (let j = 0; j < row.length; j++) {
            if (row[j] === EMPTY && !(i === 0 && j === 0))
                freeCoordinates.push([j, i]);
        }
    }

    // Find question coordinats
    freeCoordinates = shuffle(freeCoordinates);
    const questionCoordinates = freeCoordinates.splice(0, numQuestions);

    let mapDescriptor = JSON.stringify(mazeWithSurroundings);

    return {
        mapDescriptor,
        questionCoordinates,
    };
}

export function generateTiles(mapDescriptor, questions) {
    let maze = JSON.parse(mapDescriptor);
    for (let i = 0; i < questions.length; i++) {
        const {
            x,
            y,
        } = questions[i].coordinates;
        maze[y][x] = QUESTION;
    }

    let tiles = [];
    for (let i = 0; i < maze.length; i++) {
        let row = maze[i];
        let tileRow = [];

        for (let j = 0; j < row.length; j++) {
            let tile = null;

            switch (row[j]) {
                case EMPTY:
                    tile = TILE_MAPPING.grass;
                    break;

                case WALL:
                    tile = Math.random() <= 0.75 ? TILE_MAPPING.tree: TILE_MAPPING.rock;
                    break;

                case GOAL:
                    tile = TILE_MAPPING.house;
                    break;

                case QUESTION:
                    tile = Math.floor(Math.random() * 9) + 7;
                    break;
                
                default:
                    tile = TILE_MAPPING.grass;
                    break;
            }

            tileRow.push(tile);
        }

        tiles.push(tileRow);
    }

    return tiles;
}