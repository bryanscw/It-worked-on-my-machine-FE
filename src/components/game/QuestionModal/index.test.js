import React from 'react'
import { cleanup } from '@testing-library/react'
import QuestionForm from './index.js';
import { renderWithRedux } from "../../../utils/tests";

afterEach(cleanup);

it('should take a snapshot', () => {
    const { asFragment } = renderWithRedux(
        <QuestionForm
            levelId={1}
            onClose={() => {}}
            question={{
                id: 1,
                questionText: "1 + 1",
                options: {
                    "0": 1,
                    "1": 2,
                }
            }}
            isVisible={true}
            addPenalty={() => {}}
            />
    );
    expect(asFragment()).toMatchSnapshot();
});
