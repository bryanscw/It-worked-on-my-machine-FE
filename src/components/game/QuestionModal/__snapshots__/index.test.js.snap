// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="modal fade"
    data-testid="modal"
    style="padding-right: 0px;"
    tabindex="-1"
  >
    <div
      class="modal-dialog"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <h5
            class="modal-title"
          >
            1 + 1
          </h5>
          <button
            class="close"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              Ã—
            </span>
          </button>
        </div>
        <div
          class="modal-body"
        >
          <form
            class="form"
          >
            <p>
              <strong>
                Options
              </strong>
            </p>
            <div
              class="form-check"
            >
              <input
                class="form-check-input"
                id="option-0"
                name="radio"
                type="radio"
              />
              <label
                class="form-check-label"
                for="option-0"
              >
                1
              </label>
            </div>
            <div
              class="form-check"
            >
              <input
                class="form-check-input"
                id="option-1"
                name="radio"
                type="radio"
              />
              <label
                class="form-check-label"
                for="option-1"
              >
                2
              </label>
            </div>
            <div
              class="form-group mt-4"
            >
              <button
                class="btn btn-primary"
                type="submit"
              >
                Submit
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
